Add copy constructor to edmNew::DetSetVector

makortel opened the pull request:
This PR adds a ~~way to make copy of~~ copy constructor to `edmNew::DetSetVector`. Such a functionality is needed in the GPU pixel tracking work. Because the change implies recompiling ~half of the universe, I submit it here already now so that we could benefit getting it to a CMSSW pre-release (thus avoiding the aforementioned private recompilation).

~~The "copying" is implemented as the following pattern~~
~~`edmNew::DetSetVector<T> copy;`~~
~~`copy.replaceFrom(original);`~~
~~to allow putting the the `copy` to an `std::unique_ptr<T>`, and still keeping the copy constructor and assignment operator disabled (to prevent accidental copies).~~

Tested in 10_2_0_pre3, no changes expected.

@fwyzard @felicepantaleo @VinInn

makortel committed:
Implement copy constructor of edmNew::DetSetVector

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23333/4835

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

DataFormats/Common


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@rovere, @wddgit this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
So you have essentially made a strangely named assignment operator?

makortel commented:
Essentially yes. Motivation for not enabling the assignment operator (or copy constructor) is to (continue to) prevent accidental copies (e.g.`&` in `const auto foo = *fooHandle`). Real need for a copy should be rare.

Dr15Jones commented:
I'm afraid I'm not convinced by that argument. I believe if we want to do copies then we should use the standard C++ syntax for such.

makortel commented:
But we want to do copies only in very specific circumstances (that may go away in the mid-to-far future). Providing standard copy constructor and assignment operator would allow misuse and bugs in the "performance regression" sense.

Dr15Jones commented:
@makortel we have the same problems with ever std container we use now. Given that the standard operations are not used anywhere in the code at present, putting them in will not cause a regression in our present code base.

fwyzard commented:
in fact, it might make sense to `delete` the `std::string` copy constructor, and see how many copies we could get rid of...

fwyzard commented:
@Dr15Jones 
> Given that [these] are not used anywhere in the code at present, putting them in will not cause a regression in our present code base.

can I quote this as a justification to enable GCC's Concepts in the production architecture ?

cmsbuild commented:
+1
Tested at: 3d8b5aab28ba3fcc1ad321b76fc9f2ef129c87b0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23333/28192/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23333/28192/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2901706
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2901515
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 128 log files, 14 edm output root files, 31 DQM output files

makortel commented:
@Dr15Jones For my specific case a copy-constructor would be sufficient. Would it be acceptable to leave the assignment operator deleted? (I know it goes against the general recommendations)

Dr15Jones commented:
@makortel I'd be OK with that as a compromise

makortel committed:
Fix move constructor

cmsbuild commented:
The code-checks are being triggered in jenkins.

makortel commented:
Replaced the `replaceFrom()` with a copy constructor (squashed the commits so no trace if `replaceFrom()` is left in the history).

I also noticed a bug (leading to assert) in the move constructor, so I fixed that as well (meaning that before we did not have actively run code exercising the move constructor).

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-23333/4857

cmsbuild commented:
Pull request #23333 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 353a845d58cf1fd2fd117ffb89a6d3a1f2ae48ed
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23333/28221/summary.html

cmsbuild commented:
Comparison job queued.

Dr15Jones commented on DataFormats/Common/interface/DetSetVectorNew.h:
```diff
@@ -427,9 +440,19 @@ namespace edmNew {
+
     // default or delete is the same...
     DetSetVector& operator=(const DetSetVector&) = delete;
-    DetSetVector(const DetSetVector&) = delete;
```
Would allowing the default instead of hand making the constructor be enough? You could still have the comment.

makortel replied:
No, because the base class (`DetSetVectorTrans`) has two `std::atomic` members. There are comments for `DetSetVectorTrans`, I can copy them here as well if you want.

makortel replied:
Spoke too soon, of course the copy constructor here can be defaulted (as are the moves).

makortel committed:
Use '=default' for DetSetVector

Dr15Jones commented:
Please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
+1
Tested at: 79a197feae29f415a2f1a4f3b487a48f01361ea7
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23333/28232/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23333/28232/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2902474
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2902282
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 128 log files, 14 edm output root files, 31 DQM output files
